### PR TITLE
New functionality relating to proposals, events, voting machines and transactions

### PR DESCRIPTION
**Pull request by dkent600**
Resolves: #248
Resolves: #116
Resolves: #72 
Resolves: #253 
Resolves: #250 

New functionality relating to proposals, events, voting machines and transactions.  Adds support for estimating gas cost of each transaction.  Adds a few utility functions for working with tokens and their balances.

**Voting Machines**
* Adds common `IntVoteInterfaceWrapper` so that all voting machines may be treated agnostically.
* adds `getVotingMachineAddress` and `getVotingMachine` methods to each proposal-generating scheme wrapper (returning the `IntVoteInterface`).

**Proposals**
* Adds methods to proposal-generating scheme wrappers for watching/getting voteable and executed proposals
* provides rich information, including the associated voting machine, with each voteable proposal event

**Web3 Events**
* New API that facilitates transforming web3 event information into more rich "Entities for Web3 Events", including the ability to chain the events
* You can call `get` without a callback function
* You can call `subscribe` to watch web3 events as Pub/Sub events

**Transactions**
* All transaction-generation wrapper functions now return the sent (unmined) transaction hash instead of the mined transaction receipt.  `ArcTransactionResult` thus adds new methods for getting/watching the mined or confirmed transaction receipt.
* The TransactionService now publishes four events instead of one for each transaction, tracking the transaction's lifecycle:  "kickoff", "sent", "mined", "confirmed".  (Confirmation state depends on a globally-configurable block depth)
* `TransactionService` provides new methods for getting and watching mined and confirmed transactions, and for getting the current transaction depth. 
* Fixes issues with publishing transaction tracking events in the context of concurrent asynchronous transaction-generatimg operations

**Gas cost estimation**
* When the global config setting "estimateGas" is true, every transaction will be sent with a pre-estimated gas limit computed as:  `Math.max(Math.min((await func.estimateGas(...params)), maxGasLimit), 21000)`
* The default for the global config setting "estimateGas" is false.  This feature is best considered "experimental", as it hasn't been thoroughly tested on Kovan and mainnet.
* If the estimateGas function crashes, we fallback to the default gas limit.

**Breaking**
* All transaction-generation wrapper functions now return as `tx` the sent (unmined) transaction hash instead of the mined transaction receipt.
* `DAO.getDaos` now returns `EventFetcherFactory`
* `ContributionReward.getDaoProposals` is replaced by `getVotableProposals` and `getExecutedProposals`
* `EventFetcherFactory.watch` now invokes the callback once for each event.  There is no array given to the callback.
* The "TxReceipts" prefix on `TransactionService` events is changed to "TxTracking"
* `TransactionService` events changes:
    * Four `TransactionService` events are published instead of one (".kickoff", ".sent", ".mined", ".confirmed").
    * The event payload `tx` property is now the tx hash, not the receipt.  The receipt, for mined and confirmed transactions, is contained in 'txReceipt'.
    * For backwards compatibility, subscribe to `["TxTracking.theFunctionName.kickoff", "TxTracking.theFunctionName.mined"]` and use the payload 'txReceipt' property to get the transaction receipt.
* removed `DAO.getGenesisDao`
* `ProposeToRemoveGlobalConstraintParams.globalConstraint` is renamed to `ProposeToRemoveGlobalConstraintParams.globalConstraintAddress`
* The `ArcTransactionProposalResult.proposalId` property is replaced by `getProposalIdFromMinedTx()`.  
* The `ArcTransactionResult.getValueFromTx` function is replaced by `getValueFromMinedTx()`
* `TransactionService.generateInvocationKey` no longer takes an argument and now returns an integer `number` instead of a `symbol`.
* `Utils.getValueFromLogs` has moved to `TransactionService.getValueFromLogs`

**Known Issues**
* no way to obtain get parameters on global constraints when using UController.  Is a bit awkward with Controller.
* working with proposals is limited in `VestedAgreement`
* no events on `IntVoteInterface`

These will be addressed when we eventually upgrade to the next version of Arc.

### CODE REVIEW

File: lib/contractWrapperBase.ts [hunk 1]
```diff
@@ -341,128 +263,129 @@ export interface TransactionReceiptTruffle {
   /**
    * address of the transaction
    */
-  tx: Address;
+  // tx: Address;
```
Comment (tibetsprague) [inline lib/contractWrapperBase.ts]: why not just remove if not needed?

File: lib/contractWrapperBase.ts [hunk 2]
```diff
@@ -341,128 +263,129 @@ export interface TransactionReceiptTruffle {
...
-  public tx: TransactionReceiptTruffle;
+  public async getTxMined(): Promise<TransactionReceiptTruffle | null> {
+    const tx = await TransactionService.getMinedTransaction(this.tx);
+    return TransactionService.toTxTruffle(tx, this.contract);
+  }
 
-  constructor(tx: TransactionReceiptTruffle) {
-    this.tx = tx;
+  /**
+   * Returns promise of a confirmed transaction if it already exists,
```
Comment (tibetsprague) [inline lib/contractWrapperBase.ts]: meaning only return the transaction if it is already confirmed?

**Comment by dev-matan-tsuberi** [comment]
Trying to run `npm start docs.build.andPreview` receiving `No such file or directory: '/home/matan/projects/daostack/arc.js/docs/api/interfaces/EventContext.md'`

**Comment by dkent600** [comment]
@dev-matan-tsuberi 

> Trying to run npm start docs.build.andPreview …..

Yeah, I got interrupted int he middle of fixing that.  Just committed the fix.  Try again.  FYI, I still need to review the documentation and likely will be making updates soon, depending on other priorities.  But feel free to make suggestions.

File: lib/commonTypes.ts [hunk 3]
```diff
@@ -58,26 +47,26 @@ export enum SchemePermissions {
  * Note that '1' is always assigned to a scheme by the Controller when the
  * scheme is registered with the controller.
  */
-export enum DefaultSchemePermissions {
```
[the remark this answers is not in the document]
Comment (dkent600) [reply lib/commonTypes.ts]: It resulted in much cleaner code.

File: lib/contractWrapperBase.ts (hunk 1 above)
Comment (dkent600) [reply lib/contractWrapperBase.ts]: removed

File: lib/commonTypes.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (dev-matan-tsuberi) [reply lib/commonTypes.ts]: seems the same for me but sure...

File: lib/contractWrapperBase.ts (hunk 2 above)
Comment (dkent600) [reply lib/contractWrapperBase.ts]: I have clarified these comments (I hope)

File: lib/pubSubEventService.ts [hunk 4]
```diff
@@ -117,6 +182,7 @@ export class EventSubscription implements IEventSubscription {
   public constructor(private key: EventSubscriptionKey) {
   }
   public unsubscribe(): void {
-    PubSub.unsubscribe(this.key);
+    // timeout to allow lingering events to be handled before unsubscribing
+    setTimeout(() => { PubSub.unsubscribe(this.key); }, 0);
```
[the remark this answers is not in the document]
Comment (dkent600) [reply lib/pubSubEventService.ts]: Interesting, what are the advantages?  Are they worth spending time on now?

**Comment by dev-matan-tsuberi** [comment]
Why was `getProposals` removed on `ContributionRewardWrapper`?

**Comment by dkent600** [comment]
@dev-matan-tsuberi 

> Why was getDaoProposals removed on ContributionRewardWrapper and what is the alternative?

This is described in the PR description above:

> ContributionReward.getDaoProposals is replaced by getVotableProposals and getExecutedProposals

File: lib/pubSubEventService.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dev-matan-tsuberi) [reply lib/pubSubEventService.ts]: mainly cross-browser support.

[the remark this answers is not in the document]
Comment (dkent600) [reply lib/pubSubEventService.ts]: Looks at the asap code, it appears to only work in the context of Node.js.

**Comment by dev-matan-tsuberi** [comment]
Here's the branch I'm working on: https://github.com/daostack/alchemy/tree/IProposalsSquashed

File: lib/pubSubEventService.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dev-matan-tsuberi) [reply lib/pubSubEventService.ts]: https://github.com/kriskowal/asap#compatibility

> ASAP is tested on Node.js v0.10 and in a broad spectrum of web browsers.

File: lib/avatarService.ts [hunk 5]
```diff
@@ -29,7 +29,7 @@ export class AvatarService {
   }
 
   /**
-   * Returns promise of the Avatar TruffleContract.
+   * Returns promise of the Avatar Truffle contract wrapper.
```
Comment (orenyodfat) [inline lib/avatarService.ts]: Why arc.js needs the  truffle contracts ?
if there is a good reason  ..please document that.
Where is the dependency point of arc.js on truffle ?

Comment (dkent600) [reply lib/avatarService.ts]: @orenyodfat 

> Why arc.js needs the truffle contracts ?

1)  Arc.js uses them intternally
2) They are made readily accessible to users to facilitate operations that Arc.js does wrap, either now or perhaps ever.
3) The most important dependencies on truffle might be the `ContractWrapperBaseFactory` class, the `hydrate` methods in `ContractWrapperBase` and all of the wrapper methods that call Arc contract methods.

At your suggestion, I added some docs to explain how to obtain truffle contract and web3.

**Comment by dkent600** [comment]
@dev-matan-tsuberi 

> Add a failed event along with kickoff, sent, etc.. for any errors that may occur during the transaction lifecycle.

This is done.

**Pull request merged**